[GEOS-11284] Promote community module "datadir catalog loader" to core

groldan opened the pull request:
[![GEOS-11284](https://badgen.net/badge/JIRA/GEOS-11284/0052CC)](https://osgeo-org.atlassian.net/browse/GEOS-11284) [<img width="16" alt="Powered by Pull Request Badge" src="https://user-images.githubusercontent.com/1393946/111216524-d2bb8e00-85d4-11eb-821b-ed4c00989c02.png">](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

`DataDirectoryGeoServerLoader` is a faster alternative to `DefaultGeoServerLoader`, especially over network drives like NFS shares.
 
It was created in 2022 for the [GeoServer Cloud](https://github.com/geoserver/geoserver-cloud) project and then incubated in GeoServer as a community module.

Since then it's been used in several GeoServer Cloud production deployments, and now it's being moved to GeoServer's `main` module as the new default catalog and config loader for data directories.

This loader parallelizes both I/O calls and parsing of Catalog and Config info objects,
minimizing I/O calls as much as possible, trying to make a single pass over the `workspaces`
 directory tree, and loading both catalog and config files in one pass.

The point is that large Catalogs contain several thousand small XML files that need to be read
and parsed, and network shares (NFS in particular) are really bad at serving lots of small files.

Though this is the default data directory loader since GeoServer 2.25, it can be disabled (falling
back to `DefaultGeoServerLoader`), through the `datadir.loader.enabled=false` or
`DATADIR_LOADER_ENABLED=false` System Property or environment variable.

The loading process is multi-threaded, and will take place in an work-stealing `Executor` whose
parallelism is determined by an heuristic resolving to the minimum between `16` and the
number of available processors as reported by `Runtime#availableProcessors()`.

The parallelism level can also be overridden through the environment variable or system
property `DATADIR_LOAD_PARALLELISM`. A value of zero or less will produce a warning and
fall back to the default value heuristic mentioned above.

<!--Include a few sentences describing the overall goals for this Pull Request-->
  
<!-- Please help our volunteers reviewing this PR by completing the following items. 
Ask in a comment if you have troubles with any of them. -->

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

<!--Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.-->

groldan commented on src/main/src/main/java/org/geoserver/catalog/impl/DefaultCatalogFacade.java:
```diff
@@ -1035,7 +1035,7 @@ public void syncTo(CatalogFacade dao) {
             // do an optimized sync
             DefaultCatalogFacade other = (DefaultCatalogFacade) dao;
-
+            Catalog catalog = other.getCatalog();
```
Note this fixes a pre-existing issue by which all loaded `StoreInfo`s and `ResourceInfo`s end up with a reference to the discarded `Catalog`, which is then never garbage collected. 
Everything kept working as expected by accident, due to the fact that `CatalogImpl.sync()` will take ownership of the discarded catalog internals (resource pool, facade's CatalogInfoLookup's, etc)

aaime reviewed (commented):
I started the review on Monday but could not complete it. Since I'm not in a position to complete it this week (some emergencies popped up) I'm leaving the few comments I've gathered so far, as a starting point, but will need to finish the work later.

If you prefer to work through the feedback in a single go, you can safely ignore the comments for the moment, hopefully I'll manage to complete this next Monday.

aaime reviewed (commented):
Added some more notes. 

I've also tested interactively with a data directory that's really a dumpster, and managed to just lock up GeoServer for good, the parallel load never completes.
Setting up ``-Ddatadir.loader.enabled=false`` instead, the classic loader reports the same errors, but manages to complete the startup without further issues (end result, a working geoserver where a few layers are not functioning).

Here is some logs and a jstack:
* [stack.txt](https://github.com/geoserver/geoserver/files/14404157/stack.txt)
* [logs.txt](https://github.com/geoserver/geoserver/files/14404189/logs.txt)

aaime commented:
I see there has been a recent push, and some test failures:

```
Error: 8,790 [ERROR] Tests run: 5, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.563 s <<< FAILURE! - in org.geoserver.config.DataDirectoryGeoServerLoaderEnablementTest
Error: 8,790 [ERROR] testDisabledWithEnvVariable(org.geoserver.config.DataDirectoryGeoServerLoaderEnablementTest)  Time elapsed: 0 s  <<< FAILURE!
java.lang.AssertionError
	at org.geoserver.config.DataDirectoryGeoServerLoaderEnablementTest.testDisabledWithEnvVariable(DataDirectoryGeoServerLoaderEnablementTest.java:61)
```

Do you think the lockups during startup are gone? Should I retry?

petersmythe commented:
Oh dear, I thought that this had already been added to 2.25.x and wrote this for the release note.  Parking it here for when the work is completed.

### datadir Catalog loader promoted to core

One such community module has just been promoted to the core GeoServer codebase. 

It was created in 2022 for the GeoServer Cloud project and then incubated in GeoServer as a community module.  It is now being moved to GeoServer's main module as the new default catalog and config loader for data directories as it is much faster that the original loader.

It can still be disabled (falling back to DefaultGeoServerLoader), through the `datadir.loader.enabled=false` or
`DATADIR_LOADER_ENABLED=false` System Property or environment variable.

Thanks to Gabriel Roldan and Niels Charlier for contributing this work.

groldan commented:
rebased on top of latest main, fixed merge conflicts, and squash-committed.
I'll add tests based on the stack traces Andrea provided and hopefully fix those issues

aaime commented:
Tested again, it survived the dirty data directory I have, but then locked up on a "many shapefile stores" one I use for tests... apparently loading services?

I see this in the logs right before the lockup:

```
19 ago 12:57:21 CONFIG [config.internal] - Loading workspace settings...
19 ago 12:57:21 CONFIG [config.internal] - Loaded 1 workspace-specific settings.
19 ago 12:57:21 CONFIG [config.internal] - Loading workspace services...
```

Trace wise I see:

```
"main" #1 prio=5 os_prio=0 cpu=6227,80ms elapsed=227,82s tid=0x00007be3f401f800 nid=0xc90b in Object.wait()  [0x00007be3fabfc000]
   java.lang.Thread.State: WAITING (on object monitor)
        at java.lang.Object.wait(java.base@11.0.24/Native Method)
        - waiting on <0x0000000628ba8f38> (a java.util.concurrent.ForkJoinTask$AdaptedCallable)
        at java.util.concurrent.ForkJoinTask.externalInterruptibleAwaitDone(java.base@11.0.24/ForkJoinTask.java:356)
        - waiting to re-lock in wait() <0x0000000628ba8f38> (a java.util.concurrent.ForkJoinTask$AdaptedCallable)
        at java.util.concurrent.ForkJoinTask.get(java.base@11.0.24/ForkJoinTask.java:1004)
        at org.geoserver.config.internal.GeoServerConfigLoader.loadGeoServer(GeoServerConfigLoader.java:73)
        at org.geoserver.config.internal.DataDirectoryLoader.loadGeoServer(DataDirectoryLoader.java:136)
        at org.geoserver.config.DataDirectoryGeoServerLoader.readConfiguration(DataDirectoryGeoServerLoader.java:204)
        at org.geoserver.config.DefaultGeoServerLoader.loadGeoServer(DefaultGeoServerLoader.java:75)
        at org.geoserver.config.GeoServerLoader.postProcessBeforeInitializationGeoServer(GeoServerLoader.java:307)
        at org.geoserver.config.GeoServerLoader.postProcessBeforeInitialization(GeoServerLoader.java:284)
        at org.geoserver.config.GeoServerLoaderProxy.postProcessBeforeInitialization(GeoServerLoaderProxy.java:62)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:440)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1796)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:620)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:336)
        at org.springframework.beans.factory.support.AbstractBeanFactory$$Lambda$267/0x000000084041e040.getObject(Unknown Source)
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
        - locked <0x000000060c9d0a00> (a java.util.concurrent.ConcurrentHashMap)
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:334)
...

"DatadirLoader-1-worker-1" #36 daemon prio=5 os_prio=0 cpu=1592,71ms elapsed=222,04s tid=0x00007be3f63ec800 nid=0xc992 in Object.wait()  [0x00007be3b0efd000]
   java.lang.Thread.State: WAITING (on object monitor)
        at java.lang.Object.wait(java.base@11.0.24/Native Method)
        - waiting on <0x0000000628074aa0> (a java.util.stream.ReduceOps$ReduceTask)
        at java.util.concurrent.ForkJoinTask.internalWait(java.base@11.0.24/ForkJoinTask.java:311)
        - waiting to re-lock in wait() <0x0000000628074aa0> (a java.util.stream.ReduceOps$ReduceTask)
        at java.util.concurrent.ForkJoinPool.awaitJoin(java.base@11.0.24/ForkJoinPool.java:1730)
        at java.util.concurrent.ForkJoinTask.doInvoke(java.base@11.0.24/ForkJoinTask.java:411)
        at java.util.concurrent.ForkJoinTask.invoke(java.base@11.0.24/ForkJoinTask.java:736)
        at java.util.stream.ReduceOps$ReduceOp.evaluateParallel(java.base@11.0.24/ReduceOps.java:919)
        at java.util.stream.AbstractPipeline.evaluate(java.base@11.0.24/AbstractPipeline.java:233)
        at java.util.stream.ReferencePipeline.reduce(java.base@11.0.24/ReferencePipeline.java:558)
        at org.geoserver.config.internal.GeoServerConfigLoader.loadServices(GeoServerConfigLoader.java:109)
        at org.geoserver.config.internal.GeoServerConfigLoader.readGeoServer(GeoServerConfigLoader.java:94)
        at org.geoserver.config.internal.GeoServerConfigLoader$$Lambda$586/0x0000000840780c40.call(Unknown Sourc
...


"DatadirLoader-1-worker-4" #39 daemon prio=5 os_prio=0 cpu=3249,38ms elapsed=222,04s tid=0x00007be264002000 nid=0xc995 waiting for monitor entry  [0x00007be3b0bfc000]
   java.lang.Thread.State: BLOCKED (on object monitor)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.getSingletonFactoryBeanForTypeCheck(AbstractAutowireCapableBeanFactory.java:1007)
        - waiting to lock <0x000000060c9d0a00> (a java.util.concurrent.ConcurrentHashMap)
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.getTypeForFactoryBean(AbstractAutowireCapableBeanFactory.java:907)
        at org.springframework.beans.factory.support.AbstractBeanFactory.isTypeMatch(AbstractBeanFactory.java:638)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.doGetBeanNamesForType(DefaultListableBeanFactory.java:583)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeanNamesForType(DefaultListableBeanFactory.java:550)
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.getBeanNamesForType(DefaultListableBeanFactory.java:536)
        at org.springframework.context.support.AbstractApplicationContext.getBeanNamesForType(AbstractApplicationContext.java:1305)
        at org.geoserver.platform.GeoServerExtensions.extensionNames(GeoServerExtensions.java:234)
        at org.geoserver.platform.GeoServerExtensions.extensions(GeoServerExtensions.java:137)
        at org.geoserver.platform.GeoServerExtensions.extensions(GeoServerExtensions.java:292)
        at org.geoserver.security.CatalogFilterAccessManager.getCatalogFilters(CatalogFilterAccessManager.java:158)
        at org.geoserver.security.CatalogFilterAccessManager.hideWorkspace(CatalogFilterAccessManager.java:130)
        at org.geoserver.security.CatalogFilterAccessManager.getAccessLimits(CatalogFilterAccessManager.java:77)
        at org.geoserver.security.SecureCatalogImpl.buildWrapperPolicy(SecureCatalogImpl.java:965)
        at org.geoserver.security.SecureCatalogImpl.buildWrapperPolicy(SecureCatalogImpl.java:911)
...
```

All other threads are idle.

The sample data directory is [here](https://www.dropbox.com/scl/fi/ip3x82pdzui1rxnvwsqmb/large_shapefiles.tar.bz2?rlkey=7uk600hlcp9z6tdbsdka86dpl&st=bkz2knvy&dl=1)

aaime commented:
Keep it, bot
